### ISSUE TITLE
How will i do a pagination or lazy loading?

### ISSUE THREAD

**Issue by arunkrishna39** [brh55/react-native-masonry#32]

**Comment by brh55** [comment]
Hey @arunkrishna39, Sorry I don't have either functionality implemented, but I do plan on implementing Lazy Loading when my current application development process is at this stage.

**Comment by srameshr** [comment]
@brh55 Eagerly waiting!

**Comment by srameshr** [comment]
@brh55 Now that masonry implements ```flatlist```, we should have this feature with ```onEndReached``` and ```ListFooterComponent``` right?

**Comment by kmcgill88** [comment]
@srameshr I believe the answer is no, not yet. @brh55's implementation of `FlatList` has not been merged yet. See #39

**Comment by srameshr** [comment]
@kmcgill88 Flatlist has been merged, but the ```FlatList``` has been wrapped by a ```ListView``` which has been deprecated.

**Comment by vvavepacket** [comment]
@srameshr do you think that if we add our own implementation in FlatList's onEndReached() and onRefresh(), we could do ininite scroll/pagination out of the box?

**Comment by srameshr** [comment]
What I am doing right now is, passing refresh component to ListView.
Can you do a FlatList rewrite of the ListView implementation under Masonry.js

**Comment by brh55** [comment]
I've done a FlatList rewrite within the experiment branch -- it works, but not efficiently. Looks to be some sort of delay with the re-render time.

**Comment by vvavepacket** [comment]
guys, I dig deep into the code, and saw that FlatList is present inside Column.js (master branch). Now, I tried to add methods like onEndReached(). I noticed it is called during rendering when images are loading.. But if I try to manually scroll down, it doesn't get called. Any idea what to do so we can have FlatList's onEndReached() method works to enable pagination?

**Comment by vvavepacket** [comment]
@brh55 can you confirm that FlatList rewrite is only for 0.3.x version?
@srameshr do you mean to create a FlatList rewrite for the master/0.4.x/latest one?

**Comment by srameshr** [comment]
@brh55 I am on the master branch and there is an big delay during inital rendering.

There are two things here:
1. These is a ListView inside Masonry.js
2. There **WAS** list view inside Column.js but now it has been converted to flat list.

I am guessing that the ListView inside Masonry.js is responsible for the terrible rendering time.

**Comment by brh55** [comment]
If ListView is responsible then it's most likely the dataSource diff function. A quick way to check this is with removing line 19 + 42 in masonry, and replacing rowHasChanged with returning true and seeing if perhaps this is where the issue may lie.

**Comment by brh55** [comment]
@srameshr Could you give that a try on your app, and see if you notice any improvements?

**Comment by srameshr** [comment]
@brh55 Nope! Its still the same. It takes 15 seconds to render just one image.

**Comment by srameshr** [comment]
@brh55 I got it. I think its because of the remote images. If I serve it with static images, it renders in a flash but when the images are remote, it takes 15 seconds at minimum.
I'm assuming its because you are calculating the dimensions of the image before rendering.
Is it possible to just render the grid with minimumHeight and then insert images after it has been downloaded and the dimensions are calculated?

**Comment by brh55** [comment]
What are your image sizes? It's deceiving on the apps and this component because they are resizing, but good role of thumb is 300kb or less.

**Comment by srameshr** [comment]
2.7 MB 😝 
Im using a resizer now!

**Comment by brh55** [comment]
@srameshr best way to kill my bandwidth 🗡

**Comment by srameshr** [comment]
@brh55 Sorry for that. Il make sure I cross verify before posting an issue.

**Comment by brh55** [comment]
@srameshr  haha no worries

**Comment by vvavepacket** [comment]
Were you guys able to implement pagination by using flatlist's onRefresh/onEndReached() methods?

**Comment by srameshr** [comment]
@vvavepacket Since the Masonry.js file still has ListView, I do something like https://github.com/brh55/react-native-masonry/issues/51.
See the last but one comment in that.

**Comment by vvavepacket** [comment]
@srameshr i see that #51 is for pull to refresh. Did you tried about doing continous loading, like when user reaches bottom of list, the flatlist will continosuly fetch new data and add it?

**Comment by srameshr** [comment]
No, I have not yet tried that. But hopefully by the time I need it, ListView gets replaced by FlatList under masonry which supports InfinteLoading and Pull to refresh.

**Comment by srameshr** [comment]
![ezgif-4-2810d897af](https://user-images.githubusercontent.com/8149408/31139636-cf706264-a88f-11e7-95bd-bdaceaca88b9.gif)
@brh55 I think the elements in the grid that are not downloaded should have an empty placeholder at least with a loading icon. See the demo attached. After I refresh the view, the GIF that is in the second position is not rendered until its completely loaded.

**Comment by kesha-antonov** [comment]
Hello guys!
How's it going?

Is it possible now to do ```loadMoreResults``` on ```onEndReached```?

**Comment by srameshr** [comment]
@brh55 so infinite loading is a feature that lot of people are waiting for. ```FlatList``` provides it out of the box since it inherits ```ScrollView``` props.

So, until the ```ListView``` remains in ```Masonry.js```, I dont think there is any way to provide this feature.

**Comment by brh55** [comment]
@srameshr I feel you, give me some time to think about a good way to implement this. I have two possible ways, but unsure of the potential and can foresee a lot of potential bugs arrising

**Comment by vvavepacket** [comment]
How are we guys 👍

**Comment by vvavepacket** [comment]
@brh55  let us know if u need to test out your idea :) we are ready testers any time :) , this feature looks promising and is in great need and will make plugin more awesome

**Comment by davidsamacoits** [comment]
Hey guys! Any news about the infinite loading/scroll feature? Is it working now?

**Comment by vvavepacket** [comment]
+1

**Comment by adieunix** [comment]
+1

**Comment by brh55** [comment]
@davidsamacoits @vvavepacket Sorry ya'll been extremely busy at work and have had to put off a lot of OSS activities. So I'm definitely looking for co-maintainers, open for PR's and we can put together an implementation on this.

**Comment by abhinav011085** [comment]
@brh55 is this done? Eagerly waiting for this feature.

**Comment by brh55** [comment]
@abhinav011085 There is lazy load, it's not actually documented officially, but it's shown in the example.
